File: Workshops/WS5.py
# Question: 1
class PiggyBank:
    def __init__(self):
        self.__balance = 0 # Private attribute
    def add_money(self, amount):
        if amount > 0:
            self.__balance += amount
    def check_balance(self):
        return self.__balance

my_bank = PiggyBank()
my_bank.add_money(100)
print(my_bank.check_balance()) # Output: 100

# Question: 2
class Car:
    def __init__(self, make):
        self.__make = make # Private attribute
    def get_make(self):
        return self.__make
    def set_make(self, new_make):
        self.__make = new_make

car = Car("Toyota")
print(car.get_make()) # Output: Toyota
car.set_make("Honda")
print(car.get_make()) # Output: Honda

# Question: 3
class Vehicle:
    def __init__(self, brand):
        self.brand = brand
    def drive(self):
        print("The vehicle is moving.")

class My_Car(Vehicle):
    def __init__(self, brand, model):
        super().__init__(brand)
        self.model = model
    def honk(self):
        print("Beep! Beep!")

my_car = My_Car("Toyota", "Corolla")
my_car.drive() # Output: The vehicle is moving.
my_car.honk() # Output: Beep! Beep!

# Question: 4
class Animal:
    def sound(self):
        print("Some generic animal sound")
class Dog(Animal):
    def sound(self):
        print("Woof!")

dog = Dog()
dog.sound() # Output: Woof!

# Question: 5
class My_Animal:
    def sound(self):
        pass
class My_Dog(My_Animal):
    def sound(self):
        return "Woof!"
class My_Cat(My_Animal):
    def sound(self):
        return "Meow!"

animals = [My_Dog(), My_Cat()]
for animal in animals:
    print(animal.sound())
# Output:
# Woof!
# Meow!

# Question: 6
class New_Animal:
    def __init__(self, name):
        self.name = name
    def make_sound(self):
        pass # We don't know what sound every animal makes
class New_Dog(New_Animal):
    def make_sound(self):
        print("Woof!")

class New_Cat(New_Animal):
    def make_sound(self):
        print("Meow!")

my_dog = New_Dog("Buddy")
my_cat = New_Cat("Whiskers")
my_dog.make_sound() # Output: Woof!
my_cat.make_sound() # Output: Meow!

File: Workshops/WS5.js
// Qustion 1
class PiggyBank {
    constructor() {
        this._balance = 0; // Private attribute using convention
    }
    addMoney(amount) {
        if (amount > 0) {
            this._balance += amount;
        }
    }
    checkBalance() {
        return this._balance;
    }
}
const myBank = new PiggyBank();
myBank.addMoney(100);
console.log(myBank.checkBalance()); // Output: 100

// Qustion 2
class Car {
    constructor(make) {
        this._make = make; // Private attribute using convention
    }
    getMake() {
        return this._make;
    }
    setMake(newMake) {
        this._make = newMake;
    }
}

const car = new Car("Toyota");
console.log(car.getMake()); // Output: Toyota
car.setMake("Honda");
console.log(car.getMake()); // Output: Honda

// Qustion 3
class Vehicle {
    constructor(brand) {
        this.brand = brand;
    }
    drive() {
        console.log("The vehicle is moving.");
    }
}
class MyCar extends Vehicle {
    constructor(brand, model) {
        super(brand);
        this.model = model;
    }
    honk() {
        console.log("Beep! Beep!");
    }
}
const myCar = new MyCar("Toyota", "Corolla");
myCar.drive(); // Output: The vehicle is moving.
myCar.honk(); // Output: Beep! Beep!

// Qustion 4
class Animal {
    sound() {
        console.log("Some generic animal sound");
    }
}
class Dog extends Animal {
    sound() {
        console.log("Woof!");
    }
}
const dog = new Dog();
dog.sound(); // Output: Woof!

// Qustion 5
class NewAnimal {
    sound() {
        // This method will be overridden in subclasses
    }
}
class TheDog extends NewAnimal {
    sound() {
        return "Woof!";
    }
}
class TheCat extends NewAnimal {
    sound() {
        return "Meow!";
    }
}
const animals = [new TheDog(), new TheCat()];
for (const animal of animals) {
    console.log(animal.sound());
}

// Qustion 6
class TheNewAnimal {
    constructor(name) {
        this.name = name;
    }
    makeSound() {
        throw new Error("This method must be overridden by a subclass");
    }
}
class TheNewDog extends TheNewAnimal {
    makeSound() {
        console.log("Woof!");
    }
}
class TheNewCat extends TheNewAnimal {
    makeSound() {
        console.log("Meow!");
    }
}
const myDog = new TheNewDog("Buddy");
const myCat = new TheNewCat("Whiskers");
myDog.makeSound(); // Output: Woof!
myCat.makeSound(); // Output: Meow!